packages/lib/passport-interface/src/PassportPopup/core.ts

```diff
@@ -0,0 +1,219 @@
+    return "Finished. Please close this window.";
+  }
+}
+
+/**
+ * Open up a Zupass popup window using proofUrl from specific PCD integrations
+ * and popupUrl, which is the route where the useZupassPopupSetup hook is being
+ * served from.
+ */
+export function openZupassPopup(
+  popupUrl: string,
+  proofUrl: string
+): Promise<Window | null> {
+  const url = `${popupUrl}?proofUrl=${encodeURIComponent(proofUrl)}`;
+  // Calling window.open from within a React hook can cause problems.
+  // The workaround is to do it asynchronously, as per:
+  // https://stackoverflow.com/questions/76944918/should-not-already-be-working-on-window-open-in-simple-react-app
+  return new Promise((resolve) =>
+    window.setTimeout(
+      () =>
+        resolve(
+          window.open(url, "_blank", "width=450,height=600,top=100,popup")
+        ),
+      0
+    )
+  );
+}
+
+/**
+ * The popup message can contain either an encoded PCD or encoded pending PCD.
+ */
+type PopupMessageResult =
+  | {
+      type: "pcd";
+      pcdStr: string;
+    }
+  | {
+      type: "pendingPcd";
+      pendingPcdStr: string;
+    }
+  | {
+      type: "aborted";
+    };
+
+/**
+ * Asynchronous function that listens for PCDs and PendingPCDs from a Zupass
+ * popup window using message passing and event listeners.
+ */
+export function receiveZupassPopupMessage(
+  signal: AbortSignal
+): Promise<PopupMessageResult> {
+  return new Promise((resolve) => {
+    const receiveMessage = (ev: MessageEvent): void => {
+      if (ev.data.encodedPCD) {
+        resolve({ type: "pcd", pcdStr: ev.data.encodedPCD });
+        window.removeEventListener("message", receiveMessage);
+      } else if (ev.data.encodedPendingPCD) {
+        resolve({
+          type: "pendingPcd",
+          pendingPcdStr: ev.data.encodedPendingPCD
+        });
+        window.removeEventListener("message", receiveMessage);
+      }
+    };
+    window.addEventListener("message", receiveMessage, {
+      // Remove this event listener automatically if abort signal fires.
+      signal
+    });
+    signal.addEventListener("abort", () => {
+      resolve({ type: "aborted" });
+    });
+  });
+}
+
+type PopupActionResult =
+  | PopupMessageResult
+  | { type: "popupClosed" }
+  | { type: "popupBlocked" };
+
+/**
+ * An asynchronous function that opens a Zupass popup window, and returns a
+ * {@link PopupActionResult} containing either:
+ * - An encoded PCD
+ * - An encoded pending PCD
+ * - An indication that the popup was closed with no result
+ * - An indication that the popup was blocked from opening
+ *
+ * By detecting popup blockers and the popup being closed without result, this
+ * offers broader functionality than calling {@link openZupassPopup} and
+ * {@link receiveZupassPopupMessage} separately. It is added as a new API to
+ * avoid breaking backwards-compatibility for existing code.
+ */
+export async function zupassPopupExecute(
+  popupUrl: string,
+  proofUrl: string
+): Promise<PopupActionResult> {
```

Comment:
could you also explain what `popupUrl` and `proofUrl` mean?

Reply:
Done